### ISSUE TITLE
Please add flutter support to lang#dart layer

### ISSUE THREAD

**Issue by shivahnshankar** [wsdjeg/SpaceVim#3618]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Flutter is a framework that utilises Dart backend. While the dart lsp server does provide autocomplete, linting etc, there are a few features that are necessary for Flutter programming like starting emulators, hot reload, hot restart, etc. Most editors like VS Code, Emacs (Spacemacs/Doom emacs), Atom etc provide this functionality.

**Describe alternatives you've considered**
Using coc.nvim for autocompletion and coc-flutter, most of these features are available. However, coc-flutter needs to be configured manually and I believe Spacevim lang#dart layer should do this by default if the user selects to add flutter rather than plain dart.

**Additional context**
Emacs has the concept of additional packages. So using the lsp-mode for dart, flutter.el is added for this additional functionality and it works great.
[https://github.com/emacs-lsp/lsp-dart](https://github.com/emacs-lsp/lsp-dart)
In doom emacs, this means that flutter is added to the configuration like so:
(dart +flutter +lsp)

**Comment by wsdjeg** [comment]
SpaceVim has a `lsp` layer, you can enabled it for dart. and use `lsp` layer together with `lang#dart` layer. The missing feature is:

e starting emulators, hot reload, hot restart, etc ? can you list all of them? I do not use flutter, I want to know the background logic of this function.

for example, what is `starting emulators`? what is the background command?

**Comment by shivahnshankar** [comment]
If you look at the [equivalent](https://github.com/emacs-lsp/lsp-dart) in Emacs, most of these features are mentioned and are available in Spacemacs / Doom Emacs once dart + flutter + lsp are enabled.

Flutter provides a UI framework on top of Dart. When you develop Flutter apps, you typically either start an emulator or simulator, which shows how the app would look on a virtual mobile device. As you make changes, through hot reload, changes are reflected almost instantaneously in the emulator. 

Most flutter commands are accessed from the Flutter SDK though the flutter command with arguments. Here's a [good article](https://medium.com/flutter-community/flutter-and-the-command-line-a-love-story-a3648ef2411) on this. So for example.

`flutter doctor` Check whether there are any problems with Flutter installation
`flutter devices` List all devices (emulators and simulators) 
`flutter pub get` Get all dependencies in pubspec.yaml in project root dir
`flutter run` Run the app. You can also select the device id from the flutter devices command here. Once run, press 'r' for hot reload and 'R' for hot restart

So in summary:

1. Add flutter sdk (https://flutter.dev/docs/get-started/install) as a dependency for flutter apart from dart

1. If the autocomplete option Is set to COC, then there is coc-flutter (and coc-flutter-tools) extension that provide this functionality already. The lang/dart layer can be updated to include them.

1. There is [https://github.com/thosakwe/vim-flutter](https://github.com/thosakwe/vim-flutter) which I haven't tried but can be used with dart-vim-plugin, the default that lang/dart layer uses

1. Of course keybindings need to be added as well.
Hope this helps. Will be happy to provide additional details if required.

**Comment by wsdjeg** [comment]
I can not setup flutter evn in my os, I install flutter via `scoop install flutter`, it has downlowd the flutter and android-sdk, but `flutter doctor` shows:

```
[Running] flutter doctor

--------------------
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 1.22.0, on Microsoft Windows [Version 6.1.7601], locale zh-CN)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    X Flutter requires Android SDK 29 and the Android BuildTools 28.0.3
      To update the Android SDK visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions.
[!] Android Studio (not installed)
[!] Connected device
    ! No devices available

! Doctor found issues in 3 categories.

[Done] exited with code=0 in 3.755066 seconds
```

**Comment by wsdjeg** [comment]
fixed by:

```
sdkmanager --install platforms;android-29
sdkmanager --install build-tools;28.0.3
```

**Comment by wsdjeg** [comment]
I have read the code of https://github.com/thosakwe/vim-flutter/, I think the logic is very simple, we can implement this feature based on spacevim's api.

**Comment by shivahnshankar** [comment]
Great! Looking forward...and would be happy to test/document/verify in any way possible.
Thanks

**Comment by wsdjeg** [comment]
In #3876, I have added following key bindings:

### Flutter integration

When edit dart file, the following key bindings are available for running flutter commands.

| Key bindings | Descriptions                     |
| ------------ | -------------------------------- |
| `SPC l f D`  | Display flutter doctor result    |
| `SPC l f E`  | Launch to a flutter emulators id |
| `SPC l f d`  | Display flutter devices result   |
| `SPC l f e`  | Display flutter emulators list   |
| `SPC l f l`  | Reload flutter app               |
| `SPC l f r`  | Run `flutter run` command        |
| `SPC l f s`  | Restart flutter app              |


the issue is, I can not start an emulators.

`SPC l f e` displays:

```
[Running] flutter emulators

--------------------
1 available emulator:

xyz • xyz • Google • android

To run an emulator, run 'flutter emulators --launch <emulator id>'.
To create a new emulator, run 'flutter emulators --create [--name xyz]'.

You can find more information on managing emulators at the links below:
  https://developer.android.com/studio/run/managing-avds
  https://developer.android.com/studio/command-line/avdmanager

[Done] exited with code=0 in 1.715084 seconds
```

`SPC l f E`, and insert `xyz`, only warning message shown:

```
The Android emulator exited with code 1 during startup
Android emulator stderr:
Address these issues and try again.
```

I do not see any usefull message, so I can not debug it.
